### ISSUE TITLE
Harden DiscoverModelMetadata: fix hasPooledOutput detection, scope MaxTokenLength fallback, extract helper

### ISSUE THREAD

**Issue by luisquintanilla** [luisquintanilla/mlnet-text-inference-custom-transforms#28]
## Summary

`DiscoverModelMetadata()` in `OnnxTextModelScorerEstimator.cs` has several robustness issues identified during PR #26 code review. None cause failures with currently known models, but they represent latent bugs that could surface with non-standard ONNX exports.

Related: #25 (the fix that surfaced these), PR #26 (merged fix)

---

## Issue 1: `hasPooledOutput` detection rejects dynamic batch dimensions (pre-existing bug)

**Lines affected:** ~187, ~200, ~210

**Current logic:**
```csharp
hasPooledOutput = !dims.Contains(-1) && dims.Length == 2;
```

**The bug:** A pooled output like `sentence_embedding` with shape `[-1, 384]` (dynamic batch, static hidden) gets `hasPooledOutput = false` because `dims.Contains(-1)` is true (the batch dim is -1). This would cause the embedding facade to attempt pooling on an already-pooled vector.

**Why it doesn't fail today:** The auto-discovery branch (line ~223) sets `hasPooledOutput = true` directly when it finds `sentence_embedding`/`pooler_output`. The buggy `!dims.Contains(-1)` check only exists in the `OutputTensorName` and `PreferredOutputNames` branches, which aren't typically used for pooled embedding outputs. But if someone sets `PreferredOutputNames = ["sentence_embedding"]`, it would incorrectly report `hasPooledOutput = false`.

**Fix:**
```csharp
// Option A: Check only that the last dim is static and rank is 2
hasPooledOutput = dims.Length == 2 && dims.Last() > 0;

// Option B: Check by known tensor names
hasPooledOutput = dims.Length == 2 && 
    (outputName == "sentence_embedding" || outputName == "pooler_output" || !dims.Contains(-1));
```

Option A is simpler and more correct  the batch dimension being dynamic is completely normal and doesn't affect whether the output is pre-pooled.

**Files to modify:** `src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs`
- Line ~187: `OutputTensorName` branch
- Line ~200: `PreferredOutputNames` branch (preferred found)
- Line ~210: `PreferredOutputNames` branch (preferred not found)

---

## Issue 2: `MaxTokenLength` fallback is semantically wrong for pooled and rank-3 outputs

**Lines affected:** ~225-226 (pooled), ~236-237 (rank-3 unpooled)

**The problem:** PR #26 added `if (hiddenDim <= 0) hiddenDim = _options.MaxTokenLength;` to all 5 branches. This is correct for **rank-2 logit outputs** (QA models) where the last dim = sequence_length. But for:

- **Pooled outputs** (`sentence_embedding`): last dim = embedding dimension (e.g., 384), NOT sequence length
- **Rank-3 unpooled outputs** (`last_hidden_state`): last dim = hidden dimension (e.g., 384), NOT sequence length

Falling back to `MaxTokenLength` (128) for these would silently produce wrong output sizes.

**Why it doesn't fail today:** No known ONNX model has a dynamic hidden dimension for `sentence_embedding` or `last_hidden_state`. The hidden dim is always a static architecture parameter (384, 768, etc.). So the fallback never triggers.

**Fix:** Replace the `MaxTokenLength` fallback with a throw in the pooled and rank-3 branches:

```csharp
// Pooled branch (line ~225):
hiddenDim = (int)outputMeta[pooledName].Dimensions.Last();
if (hiddenDim <= 0)
    throw new InvalidOperationException(
        $"Pooled ONNX output '{pooledName}' has a dynamic hidden dimension, which is not supported. " +
        $"Set OutputTensorName explicitly or use a model with a static embedding dimension.");

// Rank-3 unpooled branch (line ~236):
hiddenDim = (int)outputMeta[outputName].Dimensions.Last();
if (hiddenDim <= 0)
    throw new InvalidOperationException(
        $"ONNX output '{outputName}' has a dynamic hidden dimension, which is not supported for auto-discovery. " +
        $"Set OutputTensorName explicitly or use a model with a static hidden dimension.");
```

Keep the `MaxTokenLength` fallback ONLY in branches where rank == 2 and the output is NOT pooled (i.e., QA logit outputs).

**Files to modify:** `src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs`
- Line ~225-226: Pooled auto-discovery branch  throw instead of fallback
- Line ~236-237: Rank-3 unpooled auto-discovery branch  throw instead of fallback

---

## Issue 3: Extract helper to reduce duplication (refactoring)

**Current state:** The pattern `hiddenDim = (int)dims.Last(); if (hiddenDim <= 0) hiddenDim = _options.MaxTokenLength;` is repeated in 5 branches.

**Suggested refactoring:**
```csharp
private int ResolveLastDimension(long[] dims, string tensorName, bool allowDynamicFallback)
{
    int lastDim = (int)dims.Last();
    if (lastDim > 0)
        return lastDim;
    
    if (allowDynamicFallback)
        return _options.MaxTokenLength;
    
    throw new InvalidOperationException(
        $"ONNX output '{tensorName}' has a dynamic last dimension, which is not supported for this output type.");
}
```

Then each branch calls:
```csharp
// QA logit outputs (rank 2, not pooled): allow fallback
hiddenDim = ResolveLastDimension(dims, outputName, allowDynamicFallback: true);

// Pooled/rank-3 outputs: fail fast
hiddenDim = ResolveLastDimension(dims, outputName, allowDynamicFallback: false);
```

This combines the fix for Issue 2 with the deduplication.

**Files to modify:** `src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs`

---

### Test Issue 1 (hasPooledOutput)
1. Create a mock scenario where `PreferredOutputNames = ["sentence_embedding"]` points to a tensor with shape `[-1, 384]`
2. Verify `hasPooledOutput` is `true` (currently would be `false`)

### Test Issue 2 (fallback semantics)
1. Create a mock scenario with a pooled output tensor where last dim is -1
2. Verify it throws `InvalidOperationException` instead of silently using `MaxTokenLength`
3. Verify QA logit outputs (rank-2, both dims dynamic) still correctly fall back to `MaxTokenLength`

### Regression tests
- Existing embedding models (all-MiniLM-L6-v2): no behavior change
- QA models (minilm-uncased-squad2): still works after Issue 2 fix (rank-2 logits still get fallback)
- Classification/NER/Reranking: no behavior change (static last dim, fallback never triggers)

---

## Acceptance Criteria

- [ ] `hasPooledOutput` correctly identifies pooled outputs with dynamic batch dimension `[-1, hidden]`
- [ ] `MaxTokenLength` fallback only applies to rank-2 non-pooled outputs (QA logits)
- [ ] Pooled and rank-3 outputs with dynamic hidden dim throw clear errors instead of silently using wrong values
- [ ] Helper method reduces code duplication in `DiscoverModelMetadata()`
- [ ] All existing model types (embedding, classification, NER, reranking, QA) continue to work
